lib/CompilerImpl/ClangLibCompiler.cpp

```diff
@@ -287,26 +286,57 @@ ClangLibCompiler::runOptimizer(const std::filesystem::path &src_IR,
   llvm::Triple moduleTriple(module->getTargetTriple());
   std::string optCPUStr, optFeaturesStr;
   llvm::TargetMachine *optTMachine = nullptr;
+
+  // All the functions contained in the else will cause the program to segfault if executed
+#if LLVM_VERSION_MAJOR >= 18
+  const llvm::TargetOptions Options = llvm::TargetOptions();
+  optCPUStr = sys::getHostCPUName().str();
+  llvm::StringMap<bool> features = llvm::sys::getHostCPUFeatures();
+#else
   const llvm::TargetOptions Options =
       llvm::codegen::InitTargetOptionsFromCodeGenFlags(moduleTriple);
-  // llvm static helper function
-  std::string lookupError;
-  if (moduleTriple.getArch()) {
-    optCPUStr = llvm::codegen::getCPUStr(); // llvm static helper function
-    optFeaturesStr =
-        llvm::codegen::getFeaturesStr(); // llvm static helper function
-  }
+  std::string optCPUStr = llvm::codegen::getCPUStr();
 
+  llvm::StringMap<bool> features;
+  llvm::sys::getHostCPUFeatures(features);
+  optFeaturesStr =
+         llvm::codegen::getFeaturesStr(); // llvm static helper function
+#endif
+   // llvm static helper function
+   std::string lookupError;
+   if (moduleTriple.getArch()) {
+     optCPUStr = sys::getHostCPUName().str();
+    std::string featuresStr;
+    for (const auto &feature : features) {
+      if (feature.second) {
+        if (!featuresStr.empty())
+          featuresStr += ",";
+        featuresStr += feature.first().str();
+      }
+    }
+    optFeaturesStr = featuresStr;
+  }
+#if LLVM_VERSION_MAJOR >= 18
+  const llvm::Target *TheTarget = llvm::TargetRegistry::lookupTarget(
+      "", moduleTriple, lookupError);
+#else
   const llvm::Target *TheTarget = llvm::TargetRegistry::lookupTarget(
       llvm::codegen::getMArch(), moduleTriple, lookupError);
+#endif
   // Some modules don't specify a triple, and this is okay.
   if (!TheTarget) {
     optTMachine = nullptr;
   } else {
+#if LLVM_VERSION_MAJOR < 18
+    std::optional<Reloc::Model> reloc = std::make_optional<Reloc::Model>(llvm::codegen::getRelocModel());
+#else
+    std::optional<Reloc::Model> reloc = std::make_optional<Reloc::Model>(Reloc::Model::Static);
+#endif
+    std::optional<CodeModel::Model> code_model = std::make_optional<CodeModel::Model>(CodeModel::Model::Small);
     optTMachine = TheTarget->createTargetMachine(
         moduleTriple.getTriple(), optCPUStr, optFeaturesStr, Options,
-        llvm::codegen::getRelocModel(),
-        llvm::CodeModel::Small, // llvm::codegen::getCodeModel returns zero
+        reloc,
+        code_model, // llvm::codegen::getCodeModel returns zero
```

Comment:
I think this comment needs to be moved to the initialisation

Reply:
do you mean the beginning of the function? i just kept the old comment since the behavior is the same

Reply:
I mean to the initialisation of the `code_model` variable you just declared.